src/main/java/org/opentripplanner/middleware/tripMonitor/jobs/CheckMonitoredTrip.java

```diff
@@ -185,21 +188,58 @@ public TripMonitorNotification checkTripForNewAlerts() {
-            return TripMonitorNotification.createDelayNotification(departureDelayInMinutes, trip.departureVarianceMinutesThreshold, DEPARTURE_DELAY);
+    /**
+     * Checks whether the trip is beginning or ending at a time greater than the allowable variance. This will check
+     * whether the departure or arrival time of the whole journey has deviated to the point where the absolute value of
+     * the variance has changed more than the trip's variance. If it has, a notification is generated.
+     */
+    public TripMonitorNotification checkTripForDelay(NotificationType delayType) {
+        // the target time for trip depending on notification type
+        Date matchingItineraryTargetTime;
+
+        // the current baseline epoch millis to check if a new threshold has been met
+        long baselineItineraryTargetEpochMillis;
+
+        // the original target epoch millis that the trip would've started or ended at
+        long originalTargetTimeEpochMillis;
+
+        if (delayType == NotificationType.DEPARTURE_DELAY) {
+            matchingItineraryTargetTime = matchingItinerary.startTime;
+            baselineItineraryTargetEpochMillis = journeyState.baselineDepartureTimeEpochMillis;
+            originalTargetTimeEpochMillis = journeyState.originalDepartureTimeEpochMillis;
+        } else {
+            matchingItineraryTargetTime = matchingItinerary.endTime;
+            baselineItineraryTargetEpochMillis = journeyState.baselineArrivalTimeEpochMillis;
+            originalTargetTimeEpochMillis = journeyState.originalArrivalTimeEpochMillis;
         }
-        return null;
-    }
 
-    public TripMonitorNotification checkTripForArrivalDelay() {
-        Leg lastLeg = matchingItinerary.legs.get(matchingItinerary.legs.size() - 1);
-        long arrivalDelayInMinutes = TimeUnit.SECONDS.toMinutes(lastLeg.arrivalDelay);
-        // Last leg arrival time should not exceed variance allowed.
-        if (arrivalDelayInMinutes >= trip.arrivalVarianceMinutesThreshold) {
-            return TripMonitorNotification.createDelayNotification(arrivalDelayInMinutes, trip.arrivalVarianceMinutesThreshold, ARRIVAL_DELAY);
+        // calculate deviation from threshold in minutes
+        long deviationInMinutes = Math.abs(
+            TimeUnit.MINUTES.convert(
+                baselineItineraryTargetEpochMillis - matchingItineraryTargetTime.getTime(),
+                TimeUnit.MILLISECONDS
+            )
+        );
+
+        // check if threshold met
+        if (deviationInMinutes >= trip.departureVarianceMinutesThreshold) {
+            // threshold met, set new baseline time
+            if (delayType == NotificationType.DEPARTURE_DELAY) {
+                journeyState.baselineDepartureTimeEpochMillis = matchingItineraryTargetTime.getTime();
+            } else {
+                journeyState.baselineArrivalTimeEpochMillis = matchingItineraryTargetTime.getTime();
+            }
```

Comment:
If we are distinguishing between departure and arrival here, is the "target time" concept still necessary as an intermediate?

Reply:
I find the result of combining the logic for arrival and departure delay checks to ultimately make things harder to read. I think I would ultimately prefer two separate checks rather than a single method that accepts the NotificationType arg. Maybe there's some shared logic that could be used by both methods, but I find the multiple checks in the method and the generic language to be somewhat hard to follow.